packages/seedless-onboarding-controller/src/SeedlessOnboardingController.ts

```diff
@@ -0,0 +1,829 @@
+    authConnectionId: string;
+    groupedAuthConnectionId?: string;
+    userId: string;
+    newPassword: string;
+    oldPassword: string;
+  }) {
+    const {
+      authConnectionId,
+      groupedAuthConnectionId,
+      userId,
+      newPassword,
+      oldPassword,
+    } = params;
+
+    const { nodeAuthTokens } = this.state;
+    this.#assertIsValidNodeAuthTokens(nodeAuthTokens);
+
+    const {
+      encKey,
+      authKeyPair,
+      keyShareIndex: newKeyShareIndex,
+    } = await this.#recoverEncKey({
+      authConnectionId,
+      groupedAuthConnectionId,
+      userId,
+      password: oldPassword,
+    });
+
+    return await this.toprfClient.changeEncKey({
+      nodeAuthTokens,
+      verifier: groupedAuthConnectionId || authConnectionId,
+      verifierId: userId,
+      oldEncKey: encKey,
+      oldAuthKeyPair: authKeyPair,
+      newKeyShareIndex,
+      newPassword,
+    });
+  }
+
+  /**
+   * Encrypt and store the seed phrase backup in the metadata store.
+   *
+   * @param seedPhrase - The seed phrase to store.
+   * @param encKey - The encryption key to store.
+   * @param authKeyPair - The authentication key pair to store.
+   *
+   * @returns A promise that resolves to the success of the operation.
+   */
+  async #encryptAndStoreSeedPhraseBackup(
+    seedPhrase: Uint8Array,
+    encKey: Uint8Array,
+    authKeyPair: KeyPair,
+  ): Promise<void> {
+    this.#assertIsValidNodeAuthTokens(this.state.nodeAuthTokens);
+
+    try {
+      const seedPhraseMetadata = new SeedPhraseMetadata(seedPhrase);
+      const secretData = seedPhraseMetadata.toBytes();
+      await this.#withPersistedSeedPhraseBackupsState(async () => {
+        await this.toprfClient.addSecretDataItem({
+          encKey,
+          secretData,
+          authKeyPair,
+        });
+        return seedPhrase;
+      });
+    } catch (error) {
+      log('Error encrypting and storing seed phrase backup', error);
+      throw new Error(
+        SeedlessOnboardingControllerError.FailedToEncryptAndStoreSeedPhraseBackup,
+      );
+    }
+  }
+
+  /**
+   * Verify the password of the encrypted vault.
+   *
+   * Upon successful verification, reterieved the nodeAuthTokens, and updates the state with the restored nodeAuthTokens.
+   *
+   * @param password - The password to verify.
+   * @returns A promise that resolves to the decrypted vault data.
+   * @throws If the password is incorrect, throw 'incorrect password' error from the #encryptor.decrypt
+   */
+  async #verifyPassword(password: string): Promise<{
+    nodeAuthTokens: NodeAuthTokens;
+    toprfEncryptionKey: Uint8Array;
+    toprfAuthKeyPair: KeyPair;
+  }> {
+    return this.#withVaultLock(async () => {
+      assertIsValidPassword(password);
+
+      const encryptedVault = this.state.vault;
+      if (!encryptedVault) {
+        throw new Error(SeedlessOnboardingControllerError.VaultError);
+      }
+
+      const decryptedVaultData = await this.#vaultEncryptor.decrypt(
```

Comment:
Note that vault decryption using the password is a very costly operation as it involves deriving the encryption key from the password using an intentionally slow key derivation function. We should make sure that we only call it very intentionally.